### DISCUSSION TITLE
Record is  immutable at compile time but it's not at runtime.

### DISCUSSION THREAD

**Discussion by moh-hassan** [dotnet/csharplang#3904] [General] [answered]
Using Net 5 Rc1 and Creating a record:

```cs
public record Rectangle
{
    public int Width { get; init; }
    public int Height { get; init; }
}
```
Then instantiating it:

```cs 
var rectangle = new Rectangle (20,30);
```
Trying to change the value:

```
rectange.Width=50; //compiler error
```
Compiler raise the error:
> error CS8852: Init-only property or indexer 'Rectangle.Width' can only be assigned in an object initializer, or on 'this' or 'base' in an instance constructor or an 'init' accessor. 

That is right and insure that the record is immutable.

Using a method [like](https://stackoverflow.com/a/60356146/3142139) to test IsImmutable type give false, because in record there is no generated readonly properties.

How to check the record in c# 9 is immutable at runtime or even it's record with `init property` ?

**Comment by YairHalberstadt** [comment, marked as answer]
A record is indeed mutable at runtime. This is intentional, is it means most serializer frameworks work without updating.

It is however possible to check if a property is init by checking:

```csharplang
bool IsInitOnly(PropertyInfo propertyInfo)
{
    return propertyInfo.SetMethod.ReturnParameter.GetRequiredCustomModifiers().Any(x => x.FullName == typeof(System.Runtime.CompilerServices.IsExternalInit).FullName);
}
```

**Reply by MgSam** [reply]
They should put a convenience property in the BCL that checks this.

**Comment by moh-hassan** [comment]
Thanks @YairHalberstadt  for answer.
It returns false for both fields.

**Reply by YairHalberstadt** [reply]
Sorry I made a mistake - updated my above answer

**Comment by moh-hassan** [comment]
Now, it returns true :) for both properties.
So, I'll check all properties of record returns true.
just a question:
I want to apply this method to records only, 
how to check the type is record to selectively apply this method at runtime (reflection show record as a class)?

**Reply by YairHalberstadt** [reply]
There's no official way to check if a type is a record, however you can check if it has a private method called `<Clone>$()`. Since this is an illegal name in C#, it is almost definitely auto generated as part of a record.

**Reply by 333fred** [reply]
It is _for now_. There is no guarantee that this will remain true, and there is intentionally no way to determine whether a type is a record or not.

**Comment by moh-hassan** [comment]
Thank you @YairHalberstadt  for help and support.
It's working.

**Comment by moh-hassan** [comment]
@333fred 
Ok, I'll take your note into consideration.
